Add block spacing gap config to theme.json and add support for this CSS variable to the "flow/default" layout.

youknowriad opened the pull request:
Folks have asked for some time now for a consistent way to tweak the margin/gap between blocks. This PR tries to solve that.

**Decisions and choices made on this PR**

 - The "block gap" is a generic style that can apply to any container block no matter of its layout type (right now we have flow and flex layouts but more will come)
 - Each layout type can "consume" the gap configuration and CSS variable and apply it to its children properly. In the current PR, the "flow" (or default) layout type uses the gap config to space the blocks vertically. While not implemented yet the "flex horizontal" layout could use the gap to space the blocks horizontally.
 - No UI/block support has been added for now, but we could imagine the gap being added to the inspector controls of container blocks (regardless of their layout type) to tweak its value contextually.
 - We'll provide a default style for this property for themes with theme.json file. This is most likely going to impact themes that are trying to switch to theme.json. I think the impact is not going to be very big but it's worth mentioning. 

**Notes**

For block templates a global CSS style applied to the root of the block templates is added by default to the theme.json generated stylesheet. Unfortunately for classic templates, we can't generate that styles automatically in the frontend because the "wrapper" for post-content doesn't have a consistent className (like `.entry-content` or something). This is also the exact same reason we ask these theme-authors to add "layout" (wide/full alignments) styles on their own to the frontend. We could potentially try a solution for this separately by relying on a convention.

**Testing instructions**

 - tweak the `"spacing": { "gap": "some value" }` in your theme.json file
 - Ensure that it impacts editor and frontend properly 
 - Check that there's no specific block styles overriding this gap.

youknowriad committed:
Add a block gap spacing style to theme.json and support in flow layout

youknowriad committed:
Remove block styles conflicting with the margin config

youknowriad commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -835,6 +837,10 @@ private function get_block_styles( $style_nodes, $setting_nodes ) {
 			$block_rules .= self::to_ruleset( $selector, $declarations );
+
+			if ( self::ROOT_BLOCK_SELECTOR === $selector ) {
+				$block_rules .= ' .wp-site-blocks > * + *, body > * + * { margin-top: var( --wp-theme-block-gap ); margin-bottom: 0; }';
```
This is introducing a small bug in the frontend (extra margin at the top of block templates), I'll see how to fix it properly.

youknowriad commented on packages/block-library/src/group/theme.scss:
```diff
@@ -2,7 +2,5 @@
 		// Matches paragraph Block padding
 		padding: $block-bg-padding--v $block-bg-padding--h;
-		margin-top: 0;
-		margin-bottom: 0;
```
This was conflicting with the this consistent margin styling. @jasmussen maybe you know where in our code base we define the block margins and whether we should move these to classic.css now.

youknowriad committed:
Fix template mode block gap support

youknowriad commented:
@ellatrix This config doesn't seem to work in the site editor for some reason, not sure I understand why.

youknowriad committed:
Fix phpcs

carolinan commented:
I find this concept more confusing than simply having margin support.

jasmussen commented:
Nice, thanks for working on this. I'm a big fan of gap! (See also #32366).

As far as the instructions in this PR, this also works as advertised, however it doesn't appear to be the `gap` that I was expecting. I'm getting this:

```
.editor-styles-wrapper .edit-post-visual-editor__post-title-wrapper > * + *, .editor-styles-wrapper .block-editor-block-list__layout.is-root-container > * + * {
    margin-top: var( --wp-theme-block-gap );
    margin-bottom: 0;
}
```

Would I get true `gap` if I set my `layout` to be a flex or grid container? 

The margin-based gap as I see here is useful enough on its own, but the reason I'm longing for "true" `gap` is that it lets you stop worry about first-child/last-child margins entirely. I often compare it to Figma's auto-layout:

![autolayout](https://user-images.githubusercontent.com/1204802/127996113-f5ff9963-4b1c-4a8a-b24c-01829ee564dd.gif)

Basically with using only _padding_ and _gap_ (Figma calls it "Space between items"), you can quite easily create consistent spacing. This is especially useful in masonry galleries, where you might have to tackle gnarly nth-child rules.

youknowriad commented:
This is inspired by on some advanced layout work you can read about here https://every-layout.dev/layouts/stack/
The gap only works in flex containers and for "flex" layout, that's exactly what we'll be using, but for "flow" layout, that trick is needed and it solves all the last-child/first-child issues.

youknowriad commented:
> I find this concept more confusing than simply having margin support.

Margin is not something that will solve this issue consistently, theme authors will need to be able to target, first and last blocks... Also, it doesn't scale well to all types of layouts.

mtias commented:
The key mental shift here is that margin, conceptualized as the spacing between blocks, is something extrinsic to the block itself and better imagined as a property of the container. Individual blocks can overwrite margin values, but it should be sporadic and considered more advanced (it can be difficult to handle for a user and break visual rhythm).

I think we might want to avoid `gap` to designated it, though, and lean more strongly on the "block" property. The description should be "Set the spacing between blocks". For the property perhaps `spacing.blocks` is fine? If we want to keep gap, maybe it should be `blockGap`.

Having control over this globally and per container should simplify how you need to deal with blocks and margins, as it should handle dealing with the top / bottom blocks by default, which has been a bit painful in columns already. In the "spacing" panel we should also be clear when a block is overwriting the default gap setting.

youknowriad committed:
Fix php unit test

youknowriad committed:
Rename gap as blockGap

youknowriad commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -233,6 +234,7 @@ class WP_Theme_JSON_Gutenberg {
 		'padding-right'              => array( 'spacing', 'padding', 'right' ),
 		'padding-bottom'             => array( 'spacing', 'padding', 'bottom' ),
 		'padding-left'               => array( 'spacing', 'padding', 'left' ),
+		'--wp-theme-block-gap'       => array( 'spacing', 'blockGap' ),
```
Is this CSS variable name good enough and consistent without our theme.json naming scheme? cc @nosolosw @jorgefilipecosta

scruffian replied:
Should it be `--wp--theme--block-gap`?

oandregal replied:
That I know of, we've used these prefixes so far:

- `--wp--preset--...` => for presets
- `--wp--custom--...` => for custom properties defined by the theme
- `--wp--style--...` => for the link color (`--wp--style--color--link`, which is now deprecated)

If this is meant to become a style themes can target, I'd perhaps recommend `--wp--style--block-gap`. I'm not too opinionated other than not having `theme` in the variable name.

oandregal replied:
We also need to declare this in the client: `__EXPERIMENTAL_STYLE_PROPERTY` at `packages/blocks/src/api/constants.js` so the site editor can generate it as well.

youknowriad commented on lib/experimental-default-theme.json:
```diff
@@ -236,5 +236,8 @@
 		}
+	},
+	"styles": {
+		"spacing": { "blockGap": "30px" }
```
@jasmussen what should be the default gap used? What do we do currently in CSS and where are the rules adding the margins?

jasmussen replied:
The default block margin as defined in _variables.scss has, for the longest time, been 28px. However that value was arrived upon based on a not-grid-based 2 * 14px "block padding" which we had in the pre-G2 block UI design, so it's entirely arbitrary. 

In classic themes, that margin is now stored in classic.scss:

```
html :where(.wp-block) {
  max-width: 840px;
  margin-top: 28px;
  margin-bottom: 28px;
}
```

(I'm debugging #32659 hence the different syntax)

In block themes at the moment we don't add any top/bottom margins to `.wp-block` at all. 

I would love for us to pick a base8 or base12 value, such as 24px, 32px or 36px. Here's what 24px looks like with a 16px font size and 32px line-height:

<img width="892" alt="Screenshot 2021-08-05 at 11 13 39" src="https://user-images.githubusercontent.com/1204802/128324995-ad665c88-524f-4b28-8d66-a3cba0ff2562.png">

That would also be nice as the default font size from the browser is 16px, which would mean that a 24px gap would equate to 1.5em. Nicely rounded numbers all around.

youknowriad replied:
Got it, I updated and used 24px.

youknowriad committed:
Add block gap support to the flex layout

youknowriad commented:
I added support for the blockGap to the flex layout as well.

youknowriad committed:
Fix more php unit tests

youknowriad committed:
Use 24px default block gap

andrewserong commented:
I think this looks very cool, I like the idea of using gap to control the spacing between blocks, and the CSS variable approach looks like a clever way to do it!

> No UI/block support has been added for now,

Separately, I've been [chipping away at a PR](https://github.com/WordPress/gutenberg/pull/33991) to add in a gap block support feature in https://github.com/WordPress/gutenberg/pull/33991 as a way of trying to come up with a solution for https://github.com/WordPress/gutenberg/issues/32366. My PR is split out from the ToolsPanel / DimensionsPanel work in https://github.com/WordPress/gutenberg/pull/32392 and adds in UI / block support with a BoxControl in the inspector controls and global styles for controlling the gap. I was imagining this as a way of handling gap (and `row-gap` and `column-gap` for control over separate vertical and horizontal spacing) for blocks like Buttons, Navigation, Social Links, and possibly longer term the Gallery block. It's implemented in a similar way to margin and padding, but might not be the right way to go about it.

It looks like this PR will provide a more general solution and approach for layouts, so it looks like it might supersede my one. But I just thought I'd check in case there's anything useful in https://github.com/WordPress/gutenberg/pull/33991 that you might want to pull in here or if there's a way to combine things. I'm happy to help any way I can!

With this PR, do you imagine the layout types will be rolled out to all other container blocks (e.g. Buttons, etc) or would we have a different approach for those sorts of blocks versus, say Group? It could also be good to include the ability to separately control row vs column gap, depending on where we roll the feature out, to support different kinds of layouts.

youknowriad commented:
@andrewserong Thanks for the ping, sorry for missing your PR initially.

> With this PR, do you imagine the layout types will be rolled out to all other container blocks (e.g. Buttons, etc) or would we have a different approach for those sorts of blocks versus, say Group?

yes, that's the goal.

>  I'd check in case there's anything useful in #32571 that you might want to pull in here or if there's a way to combine things. I'm happy to help any way I can!

Definitely, I think we should just keep your PR and refactor it slightly to generate the CSS variable added in this PR instead of directly generating the style properties. I'm not sure we'd need row and column gap to be honest but maybe we can start without it and check the feedback/use-cases. My reasoning is that the difference between column/row gap might be confusing for a lot of layouts. I guess I think we should play without adding more layouts before maying a decision there.

WDYT

scruffian commented:
I wonder if this PR makes it easier to reinstate this idea: https://github.com/WordPress/gutenberg/pull/32231

The main blocker here was that the approach broke margin collapsing, but with this setting we should be able to control the margins again.

scruffian commented on lib/experimental-default-theme.json:
```diff
@@ -236,5 +236,8 @@
 		}
+	},
+	"styles": {
+		"spacing": { "blockGap": "24px" }
```
Should this be inside `settings` not `styles`?

scruffian replied:
No it shouldn't!

andrewserong commented:
> Definitely, I think we should just keep your PR and refactor it slightly to generate the CSS variable added in this PR instead of directly generating the style properties.

Thanks @youknowriad, that sounds good to me. I'm happy to refactor my PR to get it to generate the CSS variable, and test to see that it still works in the gap use cases we had in mind. I think the CSS variable approach should give us a fair bit more flexibility, too, as you've done here, since some blocks might work best controlling the gap via margins or padding instead of the specific `gap` attribute 👍

The ToolsPanel should hopefully be getting merged early this week, so that'll help unblock moving forward the work in my PR, too.

> I'm not sure we'd need row and column gap to be honest

I think we'll probably wind up finding quite a few use cases for separate horizontal / vertical (or column / row) gap values. I found a couple of good examples elsewhere on the web, particularly for lists of posts, where the gaps are non-uniform. The other main case was groups of tags or lists of links. Here's a couple:

| List of posts | List of tags |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/128660419-0b9d3a53-de6c-44c3-871b-df9d4b4e0964.png) | ![image](https://user-images.githubusercontent.com/14988353/128660492-ad7953c7-c920-4215-9d63-7c37c986a4e1.png) |

In these examples, on the Playstation blog, the list of posts has a narrow horizontal gap, but then a larger vertical gap which gives the tiles a bit more breathing room. For the tags list in Techcrunch, it's the other way around where there's more space horizontally between each tag, but they're packed in vertically which makes the space a bit tighter.

There are certainly other workarounds for that kind of spacing (@alaczek mentioned needing to add a spacer block after post content to create bigger gaps between rows when working on blog grids), but it'd be great to give users a simple control to tweak that sort of layout.

What do you think? If we're to support that, I suppose we'd want two separate CSS variables (for columns vs row gap)? I'll have a go at exploring our options this week over in #32571

oandregal commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -835,6 +837,10 @@ private function get_block_styles( $style_nodes, $setting_nodes ) {
 			$block_rules .= self::to_ruleset( $selector, $declarations );
+
+			if ( self::ROOT_BLOCK_SELECTOR === $selector ) {
+				$block_rules .= '.wp-site-blocks > * + * { margin-top: var( --wp-theme-block-gap ); margin-bottom: 0; }';
```
Can this be the default value of `$block_rules`?

A theme may or may not add a `styles` key to its `theme.json`. What should happen with themes that don't have one? In such a case, if we still want to output this ruleset, it should be part of the default `$block_rules`. However, if we don't, it's fine as it is.

youknowriad replied:
I added a default value of `gap` in the default theme.json if that's what you mean? Is that different?

oandregal replied:
I meant that, as it stands, this ruleset is not enqueued when the theme doesn't have any `style` because `style_nodes` will be empty (so the code within the loop will not be executed).

youknowriad replied:
But from my reading, it seems impossible for a theme to not have "style" since I added the default "style" to the default theme.json?

oandregal replied:
Oh, I missed that, you're right. :+1: 

Anyway, it doesn't seem necessary to check for this in every cycle of the loop if we know we always want that ruleset in the stylesheet.

youknowriad replied:
@nosolosw Another question here. This style is injected properly in the template editor (post editor) but not in the site editor, do you know why and how to fix that?

oandregal replied:
In addition to https://github.com/WordPress/gutenberg/pull/33812/files#diff-b03597cc3da199e5aa5a94950e8241135904853e7c3b82d758e42744878afae7R237 the property also needs to be declared in the client: `__EXPERIMENTAL_STYLE_PROPERTY` at `packages/blocks/src/api/constants.js` so the site editor can generate it as well.

youknowriad replied:
That's done already in the last commit, the CSS variable is generated but the style above is not included.

oandregal replied:
Ah, I see. Adding the equivalent of this line https://github.com/WordPress/gutenberg/pull/33812/files#diff-b03597cc3da199e5aa5a94950e8241135904853e7c3b82d758e42744878afae7R842 in the client at https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/editor/global-styles-renderer.js#L308 should do it.

oandregal commented on packages/edit-post/src/components/visual-editor/index.js:
```diff
@@ -233,7 +233,14 @@ export default function VisualEditor( { styles } ) {
 						) }
 						<RecursionProvider>
-							<BlockList __experimentalLayout={ layout } />
+							<BlockList
```
I'm catching up on some layout/template editor things so I may miss tons of context: the effect of this seems that we'll show the block gap everywhere (front, site editor, post editor with template editor) except in the post editor when the theme doesn't have a `theme.json` file (classic theme). Is this reading correct?

youknowriad replied:
the idea of the change here is that the editor Dom should mimic the frontend and in the case of the "post editor", when we're in "template mode" we're showing the whole template meaning the canvas wrapper need to have the `wp-site-blocks` className like the frontend. 

Incidentally, this has the effect of making sure the newly added style targeting the top level items to apply the "gap" get applied properly in the template mode.

andrewserong commented:
@youknowriad just an update on what I've been looking into for the gap block support. Now that the ToolsPanel has been merged, I've done a bit more exploration over in https://github.com/WordPress/gutenberg/pull/33991 into the UI and per-block controls for gap support, using CSS variables. I've run into two issues:

* I tried exploring separating out the columns and row values separately, which seems to work, but does result in a potentially confusing number of CSS variables. In my [PR](https://github.com/WordPress/gutenberg/pull/33991), I've wound up with three, which is a bit awkward (one for the short-hand value, and one each for row / column value). Ideally I'd like it if we can have two to maintain control over the different directions, but again I see what you mean about this being potentially confusing or tricky to keep in sync.

* It currently isn't possible for us to save the CSS variable value within inline styles in post content for users without the `unfiltered_html` capability (e.g. Author and Contributor roles). I think this is a blocker for using CSS variables at the block level for gap support. From a bit of digging, it looks like the [recent change in WP 5.8](https://core.trac.wordpress.org/changeset/50923) to kses was to allow _referencing_ a CSS variable from within inline styles, but doesn't allow users to _set_ it. I suppose one workaround could be to add the custom properties that we want to use to `gutenberg_safe_style_attrs` in `lib/compat.php`, but I imagine that could become unwieldy if we wind up with a lot of CSS variables?

The main problem with the latter issue is that even if we decided that Author and Contributor roles shouldn't have access to the UI to adjust the gap, if they went to update an existing post with block-level gap values set, then the CSS variables would be stripped from the inline styles. Do you have any ideas for what could be a good solution or workaround? I suppose one option would be for the gap _block_ support to use a different approach to the blockGap config in this PR (e.g. use `rowGap` and `columnGap` directly), but it'd be a shame to drop the CSS variables, as they allow a lot of flexibility 🤔

youknowriad commented:
@andrewserong Thanks for the feedback, two important notes here:

For the first one, I don't mind several CSS variables but it feels like something that can be introduced later if needed.
For the second one, we have two potential solutions:

 - Update the kses rules like you suggested, I honestly don't know why CSS variables are not allowed. Maybe @nosolosw or @jorgefilipecosta could shed some light on that as they had a similar issue in the past with link colors.
 - Move the style generation for the "gap" to the server. We do have some block supports that generates styles on the server: layout, duotone. We could add block gap to the mix so it won't be a big departure. The long term issue about that is that it makes the Gutenberg's output less and less portable (requiring a server callback in addition to the "save" function of the blocks). I think we kind of "accepted" that for now, but would love @mtias's point on that.

youknowriad committed:
Rename the CSS variable to --wp--style--block-gap

youknowriad committed:
Define the constant for the site editor

youknowriad committed:
Fix edit-site renderer

youknowriad commented:
So this is working well in general, but there are two remaining issues:

 - The site editor adds empty `.block-editor-block-content-overlay__overlay` divs to children blocks to allow the "overlay click" behavior. The problem is that this div breaks the ` > * + *` selector. Ideally this div shouldn't be rendered at all and replaced by `:after` or `:before` or something like that. How possible is that @jasmussen ?
 
 - I noticed an issue with the Query block, this block has a div in the backend but not in the frontend cause the block gap to not work properly either. I think this difference in terms of "divs" is harmful in more ways than just the blog gap. I thought we added the div in the frontend as well but for some reason I can't see it now in the frontend. cc @ntsekouras do you know more here ?

youknowriad commented:
To solve the first point above, I'm exploring a refactoring to the template part overlay DOM element here https://github.com/WordPress/gutenberg/pull/34012

jorgefilipecosta commented:
> Update the kses rules like you suggested, I honestly don't know why CSS variables are not allowed. Maybe @nosolosw or @jorgefilipecosta could shed some light on that as they had a similar issue in the past with link colors.


I think randomly allowing CSS variables may cause a user without unfiltered HTML capability to in fact be able to totally customize a website when the user should not. For example, if a website uses a theme that relies a lot on CSS variables a user may be able to apply lots of complex customizations and totally change the site design just by changing the CSS variables used by the theme. I guess a malicious author just by using CSS variables may for example put an overlay content above all the website and or run clickjacking attacks or even try to load SVG images with an attack vector.

Even when allowing specific variables we need to be careful for example if we allow controlling block gap variables by kses a user without permission to control the appearance of the website may be able to change the global site block gap and not just locally for a specific block.

jorgefilipecosta commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -835,6 +837,10 @@ private function get_block_styles( $style_nodes, $setting_nodes ) {
 			$block_rules .= self::to_ruleset( $selector, $declarations );
+
+			if ( self::ROOT_BLOCK_SELECTOR === $selector ) {
+				$block_rules .= '.wp-site-blocks > * + * { margin-top: var( --wp--style--block-gap ); margin-bottom: 0; }';
```
On theme.json, If I apply blockGap rule inside core/group block it seems we are not handling this case I guess we could handle it like we do on the layout "$selector > * + * { margin-top: var( --wp--style--block-gap ); margin-bottom: 0; }";?

youknowriad replied:
Not sure how to do that, would you mind helping here?

jorgefilipecosta replied:
Making as resolved because what we have already works.

andrewserong commented:
> Even when allowing specific variables we need to be careful for example if we allow controlling block gap variables by kses a user without permission to control the appearance of the website may be able to change the global site block gap and not just locally for a specific block.

Thanks for the extra context there @jorgefilipecosta, that all makes a lot of sense, much appreciated. In that case I'll try out @youknowriad's suggestion of generating the CSS variable for the individual block server-side, as that sounds like it'll help us contain it. Update: looks like it's [working pretty well](https://github.com/WordPress/gutenberg/pull/33991#issuecomment-897403821) so far in testing, will need to clean it up a bit, but server-rendering looks promising so far 😃

> For the first one, I don't mind several CSS variables but it feels like something that can be introduced later if needed.

Sounds good. I can update that PR to use the single CSS variable and we can defer splitting out the separate directions to follow-ups further down the line 👍

youknowriad commented:
> I think randomly allowing CSS variables may cause a user without unfiltered HTML capability to in fact be able to totally customize a website when the user should not.

I don't think that's a security issue personally. you can easily do the same thing without CSS variables by using `<style>` and targeting whatever selector you want.

youknowriad committed:
Render the layout styles for all container blocks

youknowriad committed:
Fix unit tests

youknowriad commented:
I think all site editor issues are now fixed and this is now behaving properly (at least for me). I'd appreciate another review.

ntsekouras commented:
>I thought we added the div in the frontend as well but for some reason I can't see it now in the frontend. cc @ntsekouras do you know more here ?

Sorry I missed the ping. The extra wrapper is also on frontend 🤔  - you can change the wrapper element though if you want to `section` etc...

youknowriad commented:
> Sorry I missed the ping. The extra wrapper is also on frontend 🤔 - you can change the wrapper element though if you want to section etc...

@ntsekouras sorry that was a false alarm, the issue was that I was using a theme with an old markup for the query block.

youknowriad committed:
Avoid remounting the container blocks after inserting the first block

jorgefilipecosta commented on lib/block-supports/layout.php:
```diff
@@ -58,12 +58,12 @@ function gutenberg_get_layout_style( $selector, $layout ) {
 		$style .= "$selector .alignright { float: right; margin-left: 2em; }";
+		$style .= "$selector > * + * { margin-top: var( --wp--style--block-gap ); margin-bottom: 0; }";
 	} elseif ( 'flex' === $layout_type ) {
 		$style  = "$selector {";
```
In my tests the block gap works inside a group block with the default layout but not with the flex layout I think it is expected for now given the PR title. But I guess we can easily also add a rule on the flex layout that uses the --wp--style--block-gap variable.

youknowriad replied:
I actually added it to the client but forgot to add it to the server :)

jorgefilipecosta commented on lib/block-supports/layout.php:
```diff
@@ -58,12 +58,12 @@ function gutenberg_get_layout_style( $selector, $layout ) {
 
 		$style .= "$selector .alignleft { float: left; margin-right: 2em; }";
 		$style .= "$selector .alignright { float: right; margin-left: 2em; }";
+		$style .= "$selector > * + * { margin-top: var( --wp--style--block-gap ); margin-bottom: 0; }";
```
Not related to this PR but a possible improvement for the future: Rules like $selector .alignleft { float: left; margin-right: 2em; }, and $selector > * + * { margin-top: var( --wp--style--block-gap ); margin-bottom: 0; } etc don't rely on any user value. If we have 10 group blocks on a page all with the same layout we repeat these rules as inline styles 10 times. Could we just have a class with these styles not relying on user values one time and then just inject the class when needed? The way the styles would only be added one time.

youknowriad replied:
I thought about this a few time it would force us to have generic `.container-flow` classNames. Worth a try I think but we should be careful about these extra classes.

jorgefilipecosta commented on lib/theme.json:
```diff
@@ -236,5 +236,8 @@
 		}
+	},
+	"styles": {
+		"spacing": { "blockGap": "24px" }
```
I'm noticing an issue if I set a gap for 60px. And I add two top-level paragraphs in a post. On the front end, the gap between paragraphs is 60px, on the template editor too but on the post editor, the gap is not the 60px.

youknowriad replied:
Which theme are you using, I can take a look.

jorgefilipecosta replied:
The tt1-blocks from theme-experiments repo.

youknowriad replied:
For me it works properly 🤔

jorgefilipecosta commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -233,6 +234,7 @@ class WP_Theme_JSON_Gutenberg {
 		'padding-right'              => array( 'spacing', 'padding', 'right' ),
 		'padding-bottom'             => array( 'spacing', 'padding', 'bottom' ),
 		'padding-left'               => array( 'spacing', 'padding', 'left' ),
+		'--wp--style--block-gap'     => array( 'spacing', 'blockGap' ),
```
The gap works a little bit differently from the other properties because there is inheritance. If I apply margin to a block the descendent blocks don't get that margin but if I apply a gap the descendants even if not direct descendants get that gap because of how CSS variables work. I guess we should probably document somewhere that gap is inherited/propagated contrary to what happens with a margin or padding.

youknowriad replied:
Good observation, I went back and forth here. I think inheriting is good here but I'm still hesitant a bit.. WDYT? Agreed about the documentation.

jorgefilipecosta replied:
I think for this case inheritance makes sense, otherwise, something like a global gap would have almost no noticeable effect. As the direct descendants in a global context are header a content area and a footer most of the time.
I guess we can document and set that expectation for the gap and see what feedback we get.

youknowriad replied:
unless I'm missing something, I didn't find a reference for all the styles that we can add the theme.json configs. Maybe we should consider creating one separately.

oandregal replied:
Perhaps within the "styles" section? https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/#styles

youknowriad replied:
Yep, I think that would be a good place for it.

jorgefilipecosta reviewed (approved):
I left some comments but things worked mostly well on my tests and I think the PR is ready to merge from the code and functionality point of view. Nice work @youknowriad 👍

youknowriad committed:
Add frontend styles for the flex layout

youknowriad committed:
Add the block gap config to the theme.json doc

andrewserong commented:
@youknowriad it looks like the update to the check to see whether or not to apply the Layout support has caused a regression for the Buttons block (the container class and style is being applied even though the block hasn't opted-in for the layout support). I've opened up a PR to revert the check in: https://github.com/WordPress/gutenberg/pull/34189 — I wasn't sure if there's a reason why the inner blocks count needed to be there, though, so feel free to update the PR or go with a different fix if there's a better approach 🙂

jasmussen commented:
I'm working on refactoring the Navigation block to use `gap` in #32367. As part of that, I'm exploring the new flex/flow layout properties and how they can/should affect the block. If you'll bear with me I have some questions that would help me along. 

- A container block gets the new "flex" layout style if it has [its layout prop in block.json set to flex](https://github.com/WordPress/gutenberg/pull/33987/files#diff-ae0c9d82f030db516daed6bf6cc267f750ca734d8638cccc32b509e715b53c71R48), is that correct?
- The navigation block already has `display: flex;` defined in its structural CSS. But I'm assuming that if we want the global styles `blockGap` property to output a correct `gap` value, we need to change the nav blocks' block.json, correct?
- Is there already a way to define the container layout in theme.json? https://github.com/WordPress/gutenberg/pull/34189#issuecomment-902528390 suggests that as soon as the block gets a `layout` property there, that container is changed from flow to flex. But in my testing, I'm still getting the `margin-top: var( --wp--style--block-gap ); margin-bottom: 0; }` rules output. 

Appreciate any pointers! CC: @ntsekouras if you have input as well, thank you.

youknowriad commented:
> A container block gets the new "flex" layout style if it has its layout prop in block.json set to flex, is that correct?

Yes, or a default layout in the "support" config. @ntsekouras was working on refactoring Buttons I think which could be a way to inform other blocks including Navigation.

> The navigation block already has display: flex; defined in its structural CSS. But I'm assuming that if we want the global styles blockGap property to output a correct gap value, we need to change the nav blocks' block.json, correct?

Once the navigation block is refactored to be a real "flex" layout, all these specific layout styles need to be removed from its stylesheet, the rest should just work. (in theory :) )

> Is there already a way to define the container layout in theme.json?

I don't think we have that yet but the use-cases for it are not clear yet. The issue you mentioned was a bug that is now closed. The block gets "flex" if the default layout is defined as "flex".

jasmussen commented:
Thank you, that's helpful! Sounds like the CSS-first refactor can move forward and be revisited further.

> I don't think we have that yet but the use-cases for it are not clear yet. The issue you mentioned was a bug that is now closed. The block gets "flex" if the default layout is defined as "flex".

I ran into wanting to create a classic header menu that featured a site logo and a navigation block left and right as part of the Header template part, and finding that the stack-margin ruleset was applied even though I went for a horizontal layout:

<img width="1211" alt="Screenshot 2021-09-01 at 10 48 45" src="https://user-images.githubusercontent.com/1204802/131641854-cd5a3530-8d0b-427b-ad34-2518e7518983.png">

But that use case is already met by #33316, since this can now just be a single navigation block inside the header. But it does support the separate conversation we were having about allowing the group block to become a horizontal flex container.

jasmussen commented:
A thing I noticed as well. We output _only_ the `.wp-site-blocks > * + *` rule which zeroes out the bottom margin and applies the top margin as a gap. We don't output a matching `.wp-site-blocks > * { margin-top: 0; margin-bottom: 0; };` to take care of the first block in the stack:

![before](https://user-images.githubusercontent.com/1204802/131647155-dd976191-fbcc-44c2-829d-b8be62369136.gif)

Since in vertical flows, those margins commonly collapse, it's been fine so far. But it seems like we should? I can open a separate ticket.

youknowriad commented:
I think that makes sense 👍

andersnoren commented:
From #33991, I got the impression that this feature is opt-in with settings.spacing.blockGap, but it seems like the `.wp-container-XXX > * + *` margin styles are output in 11.4.0 regardless of whether settings.spacing.blockGap exists or is set to false. Just double-checking if that's intentional.

youknowriad commented:
The current reasoning is that it's intentional, the style is not opt-in, the per block UI is.

andersnoren commented:
Got it. That might cause some issues for themes that have different margins for different elements (for example, larger margin for headings, tables, separators etc. than paragraphs) and at different breakpoints, since the specificity on the styles overwriting `.wp-container-XXX > * + *` needs to be very high.

youknowriad commented:
it only applies to themes with theme.json though which I think has just been release on 5.8. So hopefully not a lot of themes. The dev note is supposed to clarify that. Do you think that's a fine shipping plan?

youknowriad commented:
I'm also curious about the use-case. Would things like this solve it? 

```
styles: {
   core/heading: {
       margin: "some value" 
   }
}
```

I'm also thinking it might not be the best approach because we don't want that "margin" value when the block is inside a horizontal/flex container for instance.

andersnoren commented:
I think being able to opt out would lower the barrier for existing themes adding theme.json pretty significantly, since most themes would probably need to modify their margin styles a lot to overwrite the default gap value (in the cases where they want to overwrite them). I think there are cases where themes built with theme.json from the ground up would want to opt out as well, if many elements in the design have different margins for different breakpoints.

> I'm also curious about the use-case. Would things like this solve it?

Being able to set the gap on a block level would definitely go a long way.

I'll ask for some input in the theme review channel on Slack.

youknowriad commented:
> I think there are cases where themes built with theme.json from the ground up would want to opt out as well, if many elements in the design have different margins for different breakpoints.

The issue I have with this is that blocks shouldn't have intrinsic margins because blocks shouldn't be assuming that they are inserted in a "flow" container, they can be on a flex container, horizontal... and those margins won't make sense anymore. I don't think a theme author should have the responsibility to define all these use-cases because it will break very quickly. 

We can add a way to "not output these styles", but I fear that this is just making things worse for themes and break some blocks in different contexts. So how to solve this use-case more holistically?

MaggieCabrera commented:
> Got it. That might cause some issues for themes that have different margins for different elements (for example, larger margin for headings, tables, separators etc. than paragraphs) and at different breakpoints, since the specificity on the styles overwriting `.wp-container-XXX > * + *` needs to be very high.

I just checked this and `[class*="wp-container-"] .block-selector` is enough to override this rule. I'd want to control the margins of the blocks via theme.json [like so](https://github.com/WordPress/gutenberg/pull/33812#issuecomment-910158399). I get the problem when the block is in a flex group. My experience is that we want to establish defaults on a vertical level and then make exceptions on the horizontal one most of the time, but that might not be a good way to go about it.

andersnoren commented:
Yeah, it's a difficult balance to strike. Most theme authors will want to have some degree of control over margins though, at least on the base element level (paragraphs, lists, headings, etc.), and it can pretty quickly lead to specificity wars.

For example, the margin styles generated for the h1 element by the code below get overwritten by the new `.wp-container-XXX > * + *` styles in 11.4.0:

```
"styles": {
	"elements": {
		"h1": {
			"spacing": {
				"margin": {
					"bottom": "48px",
					"top": "64px"
				}
			},
			"typography": {
				"fontSize": "var(--wp--preset--font-size--heading-1)",
				"letterSpacing": "var(--wp--custom--letter-spacing--heading)",
				"lineHeight": "var(--wp--custom--line-height--heading)",
				"fontWeight": "var(--wp--custom--font-weight--normal)"
			}
		},
	}
}
```

So if a theme would want to have a different margin than the blockGap for h1, they would need to add CSS looking something like this to overwrite it:

```
:root h1 {
	margin-bottom: 48px;
	margin-top: 64px;
}
```

justintadlock commented:
_Should theme authors be able to opt-out?_  If this is ever a question that comes up, the answer is always:  Absolutely, 100%, yes!

The only assumption we should make is that a theme author will want to do something custom.  If we start from that position, these sorts of conversations will happen less often.

This block gap feature has created issues across the board for me.  My plan is to _work with_ it as much as possible, but I am already seeing places where it will be problematic and force specificity battles.

Doing this in a per-block level would ease some of that.  For example, my Heading elements tend to have a different top margin than Paragraph elements (this was actually the first thing I noticed that was broken).

**Edit:** On the topic of horizontal, flex, etc. containers:  those are exceptions to the rule.  For example, I target `:first-child` to remove top margins, which I think is pretty typical in design systems.  This automatically handles things like the Columns/Column blocks.  The same or similar methods would work in other types of containers.

youknowriad commented:
> Should theme authors be able to opt-out? If this is ever a question that comes up, the answer is always: Absolutely, 100%, yes!

I disagree with this take. This means that everything should be optional in WordPress and goes against the decisions not options. some things need to be options but not everything.  

After the discussion we had here, I agree that for now the block gap styles need to be opt-out because we don't have answers for all the use cases yet but I don't think it should be a rule to have an opt-out for everything personally.  For instance for structural styles, I'd rather have the themes rely on Core **always** instead of reinventing their own. Themes are here to bring personality and design but not to define what "horizontal alignment" means for instance.

> This block gap feature has created issues across the board for me. My plan is to work with it as much as possible, but I am already seeing places where it will be problematic and force specificity battles.

I'd love to learn mode about the use-cases to be able to come up with the best abstractions that solve this. If we always fallback to CSS to solve these issues, specificity issues will continue forever and adaptation as Core changes structural changes will continue to be a nightmare. Theme.json is an attempt to solve that, so we should try to reduce the need for these custom CSS (other than things that bring personality) from themes.

> Doing this in a per-block level would ease some of that. For example, my Heading elements tend to have a different top margin than Paragraph elements (this was actually the first thing I noticed that was broken).

👍 Yes that use-case has been raised above and makes sense but it's also important to understand that it only makes sense in the "flow" layout. If you insert a heading block inside a horizontal container, it shouldn't have any margins. For this particular reason, we need to be more smart about it:

 - If it's theme author's responsibility, they'll forget about these things and create broken designs
 - if it's core responsibility, we need ways to ensure these things work properly in all contexts.

MaggieCabrera commented:
I can think of other cases where you'd want a different gap per block. One example would be the "small" blocks that have post meta content such as post date or post tags, we encountered this on skatepark, where the margin bottom (or gap) for an h1 could be 5 times or more the spacing between these smaller blocks:

<img width="1235" alt="Screenshot 2021-09-01 at 18 30 47" src="https://user-images.githubusercontent.com/3593343/131709016-dfcfa4cd-4d7b-4fba-991c-b70cc4c297a4.png">

chthonic-ds commented:
Seeing the recent comments here, I was dreading how this might affect the block-based theme I'm currently working on. The effects ended up being minimal due to already adopting the "styling the context" approach every-layout.dev endorses, but at a different location: the `.post-content` level. My experience has been header/footer areas benefit less from the same rules, due to having their own micro-contexts to manage.

A few examples from my current theme where elements require different "gaps" in a flow context: 

- `<h2>` and `<h3>` following any element
- any element following a Columns block (excluding another Columns block)
- any element following a full/wide aligned element with a background colour set
- any top-level element with a background colour set following any element
- "meta" content following a heading

> _Should theme authors be able to opt-out?_ If this is ever a question that comes up, the answer is always: Absolutely, 100%, yes!

Very much agree from a theming perspective.

> For instance for structural styles, I'd rather have the themes rely on Core always instead of reinventing their own.

> If it's theme author's responsibility, they'll forget about these things and create broken designs. If it's core responsibility, we need ways to ensure these things work properly in all contexts.

I've gone back and forth since 5.0 with adding to core styles vs. 100% custom styles, and, overall, I've seen custom styles be the only way to achieve layouts I'm handed, as well as being the easiest to maintain over the long term (the other reason is, using custom styles has allowed me to use things like CSS Grid for layout _now_ rather than waiting for it to appear as a WordPress-provided option).

Working with WordPress post-5.0 has meant learning to write styles in a much more defensive and resilient way, something I see as a positive outcome. However, I do often feel idiotic for writing my own style-sheets. Perhaps it's time to send a clear signal to someone like me by removing the option to dequeue core styles?

mjsdiaz commented:
> I find this concept more confusing than simply having margin support.

I would like to be able to opt out for the theme global gap and opt-in on specific blocks.

ellenbauer commented:
> Should theme authors be able to opt-out? If this is ever a question that comes up, the answer is always: Absolutely, 100%, yes!

Yes, yes, yes!!! I think this is an absolute must, since a general rule might work for some smaller projects, but it does not leave enough choice/control for theme developers to make custom decisions.

At the moment I get the feeling in general that we might need more clarity on "what goes where" as I struggle to understand what is meant to be block 'territory' and what is theme territory and so on.
Is this just me or do others feel the same?

iandunn commented:
> we don't have answers for all the use cases yet...
> I'd love to learn mode about the use-cases to be able to come up with the best abstractions that solve this

I opened #34716 to start documenting use cases that need improvement
